crates/ty_python_semantic/resources/mdtest/binary/tuples.md

```diff
@@ -17,6 +17,35 @@ def _(x: tuple[int, str], y: tuple[None, tuple[int]]):
 
 ```py
 def _(x: tuple[int, ...], y: tuple[str, ...]):
+    reveal_type(x + x)  # revealed: tuple[int, ...]
     reveal_type(x + y)  # revealed: tuple[int | str, ...]
-    reveal_type(x + (1, 2))  # revealed: tuple[int, ...]
+    reveal_type((1, 2) + x)  # revealed: tuple[Literal[1], Literal[2], *tuple[int, ...]]
+    reveal_type(x + (3, 4))  # revealed: tuple[*tuple[int, ...], Literal[3], Literal[4]]
+    reveal_type((1, 2) + x + (3, 4))  # revealed: tuple[Literal[1], Literal[2], *tuple[int, ...], Literal[3], Literal[4]]
+    reveal_type((1, 2) + y + (3, 4) + x)  # revealed: tuple[Literal[1], Literal[2], *tuple[int | str, ...]]
+```
+
+If we do the same thing through a legacy type alias, we lose information, since this involves first
+inferring the `tuple[...]` expression as a value form. (Doing so gives a generic alias of the
+`tuple` type, which only has a single type variable in its typeshed definition. That requires us to
+summarize our knowledge of the tuple elements into a single union type.)
+
+TODO: We plan to handle legacy type aliases better, by inferring them directly as a type expression
+when they're used in such a context. That should prevent the information loss described here.
```

Comment:
The problem isn't limited to legacy type aliases: we also need to be able to preserve some knowledge of the type parameters for things like

```py
class Foo(tuple[int, int]): ...
```

which is something that's on my monthly goals, since a `NamedTuple` is really just an easy way of creating a class like ^that but that has some bells and whistles attached. And `NamedTuple`s are pretty popular!

Reply:
We now track the full element detail in the generic aliases